packages/workspace/src/browser/workspace-service.ts

```diff
@@ -57,18 +66,128 @@ export class WorkspaceService implements FrontendApplicationContribution {
 
     @postConstruct()
     protected async init(): Promise<void> {
-        const rootUri = await this.server.getRoot();
-        this._root = await this.toValidRoot(rootUri);
-        if (this._root) {
-            const uri = new URI(this._root.uri);
-            this.updateTitle(uri);
-            this.watcher.watchFileChanges(uri);
+        this._workspaceDescription = {
+            root: undefined,
+            roots: [],
+            workspaceConfigFile: undefined
+        };
+        await this.updateWorkspace();
+        this.updateTitle();
+        this.deferredWorkspaceDescription.resolve(this._workspaceDescription);
+        const configUri = this.getWorkspaceConfigFileUri();
+        if (configUri) {
+            const configUriString = configUri.toString();
+            this.watcher.onFilesChanged(changes => {
+                if (changes.some(change => change.uri.toString() === configUriString)) {
+                    this.updateWorkspace();
+                }
+            });
+        }
+    }
+
+    get root(): Promise<FileStat | undefined> {
```

Comment:
shouldn't this be now deprecated?

Reply:
And it should return the first element from the 'roots'

Reply:
Thank you for the review @svenefftinge . Really appreciated ! 

I tried what you suggested (only exposing `roots`, and the use the first element of it for `root`), and found it didn't work in all scenarios we should support. 

Think about this if we only expose `roots`: 
Step 1. We have `theia` and `vsCode` in the same workspace, with `theia` being the workspace folder. At this point, `roots = [theia, vsCode]`. No issues.
Step 2. Remove `theia` from the workspace. (Yes, removing the workspace folder is supported.)  When this is done, `roots = [vsCode]`, and there is no way for other extensions to figure out what the workspace folder is, if the user reloads theia.